pkg/alertmanager/multitenant.go

```diff
@@ -604,17 +604,16 @@ func (am *MultitenantAlertmanager) syncConfigs(cfgs map[string]alerts.AlertConfi
 
 	am.alertmanagersMtx.Lock()
 	defer am.alertmanagersMtx.Unlock()
-	for user, userAM := range am.alertmanagers {
-		if _, exists := cfgs[user]; !exists {
-			// The user alertmanager is only paused in order to retain the prometheus metrics
-			// it has reported to its registry. If a new config for this user appears, this structure
-			// will be reused.
-			level.Info(am.logger).Log("msg", "deactivating per-tenant alertmanager", "user", user)
-			userAM.Pause()
-			delete(am.cfgs, user)
-			am.multitenantMetrics.lastReloadSuccessful.DeleteLabelValues(user)
-			am.multitenantMetrics.lastReloadSuccessfulTimestamp.DeleteLabelValues(user)
-			level.Info(am.logger).Log("msg", "deactivated per-tenant alertmanager", "user", user)
+	for userID, userAM := range am.alertmanagers {
+		if _, exists := cfgs[userID]; !exists {
+			level.Info(am.logger).Log("msg", "deactivating per-tenant alertmanager", "user", userID)
+			go userAM.Stop()
```

Comment:
I would try to avoid calling `Stop()` is a goroutine. Calling it synchronously would be safer in my opinion.

The `userAM.Stop()` does basically two things:
1. Cancel contexts
2. `am.wg.Wait()`

The real blocking thing here could be the wait, so I would suggest the following:
1. Run `userAM.Stop()` synchronously (no gorouting here)
2. Remove the `am.wg.Wait()` from the `Stop()`
3. Create `StopAndWait()` which calls `Stop()` and then `am.wg.Wait()`, replacing existing calls to `Stop()` with `StopAndWait()` (except for this one)

WDYT @gotjosh @pstibrany ?

Reply:
I think this makes sense, nice catch. 

I have one question though (mostly for the benefit of my understanding), in practice the split of context cancelling and `wg.Wait()` means that we instruct the channels to close but don't wait for them to drained? If so, what happens if they never drain? I guess it would no better from before where we were just fire and forgetting about the stopping.